### PR TITLE
Enable hub & spoke transitivity via gateway VMs

### PR DESCRIPTION
**Pull request by drebes**
This enables transitivity between spokes using regional MIGs that act as transitivity gateways.

Feature gated by `enable_hub_and_spoke_transitivity` which depends on `enable_hub_and_spoke`.

Just tested locally, let me know what's needed for integration testing.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/envs/development/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/envs/development/README.md
25c25
< | enable\_hub\_and\_spoke\_transitivity | Enable Hub-and-Spoke transitivity via gateways.  Depends on `enable_hub_and_spoke_transitivity`| `bool` | `false` | no |
---
> | enable\_hub\_and\_spoke\_transitivity | Enable transitivity via gateway VMs on Hub-and-Spoke architecture. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/envs/non-production/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/envs/non-production/README.md
25c25
< | enable\_hub\_and\_spoke\_transitivity | Enable Hub-and-Spoke transitivity via gateways.  Depends on `enable_hub_and_spoke_transitivity`| `bool` | `false` | no |
---
> | enable\_hub\_and\_spoke\_transitivity | Enable transitivity via gateway VMs on Hub-and-Spoke architecture. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/envs/production/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/envs/production/README.md
25c25
< | enable\_hub\_and\_spoke\_transitivity | Enable Hub-and-Spoke transitivity via gateways.  Depends on `enable_hub_and_spoke_transitivity`| `bool` | `false` | no |
---
> | enable\_hub\_and\_spoke\_transitivity | Enable transitivity via gateway VMs on Hub-and-Spoke architecture. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/envs/shared/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/envs/shared/README.md
31c31
< | enable\_hub\_and\_spoke\_transitivity | Enable Hub-and-Spoke transitivity via gateways.  Depends on `enable_hub_and_spoke_transitivity`| `bool` | `false` | no |
---
> | enable\_hub\_and\_spoke\_transitivity | Enable transitivity via gateway VMs on Hub-and-Spoke architecture. | `bool` | `false` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/modules/base_shared_vpc/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/modules/base_shared_vpc/README.md
5a6,7
> | allow\_all\_egress\_ranges | List of network ranges to which all egress traffic will be allowed | `any` | `null` | no |
> | allow\_all\_ingress\_ranges | List of network ranges from which all ingress traffic will be allowed | `any` | `null` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/3-networks/modules/restricted_shared_vpc/README.md /tmp/tmp.RRIVahAH4M/generate_docs/workspace/3-networks/modules/restricted_shared_vpc/README.md
6a7,8
> | allow\_all\_egress\_ranges | List of network ranges to which all egress traffic will be allowed | `any` | `null` | no |
> | allow\_all\_ingress\_ranges | List of network ranges from which all ingress traffic will be allowed | `any` | `null` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
Checking for missing newline at end of file
Error: No newline at end of file ./3-networks/envs/development/variables.tf
Error: No newline at end of file ./3-networks/envs/production/variables.tf
Error: No newline at end of file ./3-networks/modules/transitivity/assets/gw.yaml
Error: No newline at end of file ./3-networks/modules/transitivity/variables.tf
Error: No newline at end of file ./3-networks/modules/transitivity/main.tf
Error: No newline at end of file ./3-networks/modules/transitivity/README.md
Error: No newline at end of file ./3-networks/modules/restricted_shared_vpc/variables.tf
Error: No newline at end of file ./3-networks/modules/base_shared_vpc/firewall.tf
Error: No newline at end of file ./3-networks/modules/base_shared_vpc/variables.tf
 ``` 
- ⚠️check_headers
 All files need a license header. Please make sure all your files include the appropriate header. A helper tool available [here](https://github.com/google/addlicense). 
 ```bash 
Checking file headers
3 files have incorrect boilerplate headers:
3-networks/envs/shared/net-hubs-transitivity.tf
3-networks/modules/transitivity/assets/gw.yaml
3-networks/modules/transitivity/main.tf
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Attempting to download /workspace/test/bundle.hcl bundle.
/tmp/bundler /workspace
Fetching Terraform 0.13.5 core package...
Local plugin directory ".plugins" found; scanning for provider binaries.
No ".plugins" directory found, skipping local provider discovery.
Creating terraform_0.13.5-bundle2021020321_linux_amd64.zip ...
All done!
Archive:  terraform_0.13.5-bundle2021020321_linux_amd64.zip
/workspace
Running terraform fmt
Running terraform validate
terraform_validate ./0-bootstrap 
Success!
 The configuration is valid.
terraform_validate ./0-bootstrap/modules/jenkins-agent 
Success!
 The configuration is valid.
terraform_validate ./1-org/envs/shared 
Success!
 The configuration is valid.
terraform_validate ./2-environments/envs/development 
Success!
 The configuration is valid.
terraform_validate ./2-environments/envs/non-production 
Success!
 The configuration is valid.
terraform_validate ./2-environments/envs/production 
Success!
 The configuration is valid.
terraform_validate ./2-environments/modules/env_baseline 
Success!
 The configuration is valid.
terraform_validate ./3-networks/envs/development 
Success!
 The configuration is valid.
terraform_validate ./3-networks/envs/non-production 
Success!
 The configuration is valid.
terraform_validate ./3-networks/envs/production 
Success!
 The configuration is valid.
terraform_validate ./3-networks/envs/shared 
Success!
 The configuration is valid.
terraform_validate ./3-networks/modules/base_shared_vpc 
Success!
 The configuration is valid.
terraform_validate ./3-networks/modules/dedicated_interconnect 
Success!
 The configuration is valid.
terraform_validate ./3-networks/modules/restricted_shared_vpc 
Success!
 The configuration is valid.
terraform_validate ./3-networks/modules/transitivity 
Success!
 The configuration is valid.
terraform_validate ./3-networks/modules/vpn-ha 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_1/development 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_1/non-production 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_1/production 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_2/development 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_2/non-production 
Success!
 The configuration is valid.
terraform_validate ./4-projects/business_unit_2/production 
Success!
 The configuration is valid.
terraform_validate ./4-projects/modules/single_project 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/bootstrap 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/envs 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/networks 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Module does not support depends_on
  on main.tf line 30, in module "non-production":
  30:   source                           = 
"../../../3-networks/envs/non-production"
Module "non-production" cannot be used with depends_on because it contains a
nested provider configuration for "google.impersonate", at
../../../3-networks/envs/non-production/providers.tf:21,10-18.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
Error: 
Module does not support depends_on
  on main.tf line 30, in module "non-production":
  30:   source                           = 
"../../../3-networks/envs/non-production"
Module "non-production" cannot be used with depends_on because it contains a
nested provider configuration for "google", at
../../../3-networks/envs/non-production/providers.tf:40,10-18.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
Error: 
Module does not support depends_on
  on main.tf line 30, in module "non-production":
  30:   source                           = 
"../../../3-networks/envs/non-production"
Module "non-production" cannot be used with depends_on because it contains a
nested provider configuration for "google-beta", at
../../../3-networks/envs/non-production/providers.tf:45,10-23.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
Error: 
Module does not support depends_on
  on main.tf line 43, in module "production":
  43:   source                           = 
"../../../3-networks/envs/production"
Module "production" cannot be used with depends_on because it contains a
nested provider configuration for "google.impersonate", at
../../../3-networks/envs/production/providers.tf:21,10-18.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
Error: 
Module does not support depends_on
  on main.tf line 43, in module "production":
  43:   source                           = 
"../../../3-networks/envs/production"
Module "production" cannot be used with depends_on because it contains a
nested provider configuration for "google", at
../../../3-networks/envs/production/providers.tf:40,10-18.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
Error: 
Module does not support depends_on
  on main.tf line 43, in module "production":
  43:   source                           = 
"../../../3-networks/envs/production"
Module "production" cannot be used with depends_on because it contains a
nested provider configuration for "google-beta", at
../../../3-networks/envs/production/providers.tf:45,10-23.
This module can be made compatible with depends_on by changing it to receive
all of its provider configurations from the calling module, by using the
"providers" argument in the calling module block.
terraform_validate ./test/fixtures/org 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/projects 
Success!
 The configuration is valid.
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

File: 3-networks/modules/transitivity/main.tf [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    balancing_mode = "CONNECTION"
+    }
+  ]
+  health_check_config = {
+    type = "tcp", check = { port = 22 }, config = {}, logging = true
+  }
+}
+
+module "vpc_routes" {
+  source     = "github.com/terraform-google-modules/cloud-foundation-fabric//modules///net-vpc?ref=v4.3.0"
```
Comment (bharathkkb) [inline 3-networks/modules/transitivity/main.tf:167]: Can we make use of the routes module instead as it is published to the registry?
https://github.com/terraform-google-modules/terraform-google-network/tree/master/modules/routes

File: 3-networks/modules/transitivity/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+TODO
```
Comment (bharathkkb) [inline 3-networks/modules/transitivity/README.md:1]: nit: todo

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR @drebes 
I have a fix https://github.com/terraform-google-modules/terraform-example-foundation/pull/323 for the `check_terraform` error [reported by the bot](https://github.com/terraform-google-modules/terraform-example-foundation/pull/322#issuecomment-772854865), but we should fix the others.
We should also update the cidr ranges being asserted in the tests [here](https://github.com/terraform-google-modules/terraform-example-foundation/blob/eea9416a19d9ee0261975e46b511a5d2d8a3433e/test/integration/networks/controls/gcp_networks.rb#L33).

File: 3-networks/modules/transitivity/main.tf (hunk 1 above)
Comment (bharathkkb) [reply 3-networks/modules/transitivity/main.tf:167]: we could also use https://github.com/terraform-google-modules/terraform-google-lb-internal which is in the registry

Comment (drebes) [reply 3-networks/modules/transitivity/main.tf:167]: I didn't find a way to set ILB as next hop with the registry's routes module.
For the internal load balancer module, it requires me to specify ports, despite it being an all_ports = true LB.

**Comment by rjerrems** [comment]
looks like there are still a few linting errors @drebes - if you would like to run this quickly locally, you can do so with this command `make docker_test_lint`

File: 3-networks/modules/transitivity/main.tf (hunk 1 above)
Comment (drebes) [reply 3-networks/modules/transitivity/main.tf:167]: Created  https://github.com/terraform-google-modules/terraform-google-lb-internal/issues/45 for the issue with firewalls in the lb-internal module.

Comment (drebes) [reply 3-networks/modules/transitivity/main.tf:167]: Created https://github.com/terraform-google-modules/terraform-google-network/issues/240 for the issue with `next_hop_ilb` routes.

File: 3-networks/modules/transitivity/main.tf [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+      targets              = [module.service_account.emails["transitivity-gw"]]
+      use_service_accounts = true
+      rules                = [{ protocol = "all", ports = null }]
+      extra_attributes     = {}
+    }
+  }
+}
+
+module "ilb_addresses" {
+  source     = "github.com/terraform-google-modules/cloud-foundation-fabric//modules/net-address?ref=v4.3.0"
```
Comment (morgante) [inline 3-networks/modules/transitivity/main.tf:128]: We should not reference Fabric modules from a different repo. They're not a stable interface.

**Review by bharathkkb — COMMENTED** [review]
Overall LGTM 
- We can add a `EXCLUDE_LINT_DIRS` env var set to `\./3-networks/modules/transitivity/assets` to work around https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/issues/870. More info [here](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/blob/16aaf234944fba1d6b7293e6b02ede7b1b2fad7d/infra/build/developer-tools/build/scripts/task_helper_functions.sh#L61). We will need to add this env var to the makefile and the cloudbuild lint pipeline.

- For integration tests we can enable `enable_hub_and_spoke_transitivity` in one of the [network env fixtures](https://github.com/terraform-google-modules/terraform-example-foundation/blob/6c06f1ec28aa79f39fb34d4a7807d30942a0406f/test/fixtures/networks/main.tf#L26) and assert that the MIGs and ILBs are present. Is there a way to test the transitivity itself?

**Comment by drebes** [comment]
@bharathkkb done with the lint tests and rebased. I want to start looking into the integration tests now but I have zero experience with kitchen. Can you provide me a link to a similar PR where tests to new network related features were added? It's not clear to me where the tests set the new variables in neither #298 nor #345

**Review by rjerrems — APPROVED** [review]
LGTM - I think we can probably add the integration tests in a subsequent PR @drebes ?

cc @bharathkkb

**Pull request merged**